### DISCUSSION TITLE
Logtalk for VSCode extension

### DISCUSSION THREAD

**Discussion by hnbeck** [LogtalkDotOrg/logtalk3#128] [Q&A]
Hi, 

developing on Win10 and with Visual Studio Code, I've trouble to get the Logtalk plugin and its Linter to work. Has anyone a running environment with Win10 or Win11? 

My problem is that the necessary configurion on LINUX and Windows are very different. On Linux, there is a `/usr/local/bin/logtalk,` which seems to have no equivalent on Win10, where the shortcut from the Start menu is `C:\Dev\swipl\bin\swipl-win.exe` (my SWI and Logtalk reside at `C: \dev`)

So my question is how to configure that Logtalk loads the file in editor and especially that the linter is running?

Cheers

Hans

**Comment by pmoura** [comment]
There's a new version of the plug-in being developed by [Jacob Friedman](https://github.com/jacobfriedman) and sponsored by [Graphstax](https://graphstax.ai). Jacob is in the process of publishing it at his GitHub account.

In this new plug-in version, when using VSC on Windows, you can set the Logtalk executable path to e.g. the SWI-Prolog executable path and then set the executable argument to load Logtalk. As an example, on my Windows 10 VM, I'm using (on my `settings.json` file):

```js
{
    "logtalk.home.path": "C:/Program Files (x86)/Logtalk",
    "logtalk.executable.path": "C:\\Program Files\\swipl\\bin\\swipl.exe",
    "logtalk.executable.arguments": [
        "-s",
        "C:\\Program Files (x86)\\Logtalk\\integration\\logtalk_swi.pl"
    ],
    "window.zoomLevel": 4,
    "logtalk.user.path": "C:/Users/Paulo Moura/Documents/Logtalk"
}
```

Note that the values of the `logtalk.home.path` and `logtalk.user.path` keys are, respectively, the values of the `LOGTALKHOME` and `LOGTALKUSER` environment variables, which are different in your case; you will need to adjust the paths accordingly. For other Prolog backends, the arguments for the `logtalk.executable.arguments` key can be taken from the Logtalk integration shortcuts that are available from the Start Menu after installing Logtalk.

**Comment by jacobfriedman** [comment]
Hello @hnbeck, please try the latest version of the [VSCode Logtalk plugin here](https://github.com/jacobfriedman/vsc-logtalk).  Since it's not updated on the marketplace at the moment, you can install the VSIX included in the repo. Any feedback is appreciated. Thank you, Paulo, for your ongoing support and help with this plugin (and, of course, thanks to [Graphstax](http://graphstax.ai) for supporting its development).

**Reply by hnbeck** [reply]
Hi Jacob,  
much better, great :) The only thing is when doing the logtalk make it misses the file vscode_message_streamer.lgt, which indeed is not part of my Logtalk installation (which should be 5.22). 

Cheers and thx

Hans

**Reply by pmoura** [reply]
Pushed that file to the current Logtalk git version. It will be in the next stable release.
https://github.com/LogtalkDotOrg/logtalk3/blob/master/coding/vscode/vscode_message_streamer.lgt

**Comment by pmoura** [comment]
Version 0.4.8 of the plug-in released.

**Comment by pmoura** [comment]
Version 0.4.9 of the plug-in released. Thanks to Jacob work, you can now Cmd/Ctrl-click on the warning or error in the embedded terminal and jump to that file and line(s) location.

**Comment by pmoura** [comment]
Development of the VSCode extension have been moved to:

https://github.com/LogtalkDotOrg/logtalk-for-vscode

Note that the extension have been renamed to "Logtalk for VSCode". We're preparing the extension for publication in the VSCode [marketplace](https://marketplace.visualstudio.com/VSCode).

**Reply by pmoura** [reply]
Until we publish in the VSCode marketplace, you can download the `.vsix` file for the latest version for manual installation (using the 'Extensions: Install from VSIX...' command from the command palette) at:

https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases

**Reply by pmoura** [reply]
Version 0.5.3 of the plug-in released fixing syntax highlight of term and arithmetic comparison operators.

**Comment by pmoura** [comment]
Version [0.6.1](https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.6.1) released. Implemented generation of diagrams and documentation in final formats (finally found a reliable solution). Be aware that settings changed. See the README.md file for details. Tested on both macOS and Windows.

**Comment by pmoura** [comment]
Version [0.8.4](https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.8.4) released. All Logtalk compiler warnings and errors are now reliably added to the "Problems" pane when compiling a file. Tested on both macOS and Windows.

**Comment by pmoura** [comment]
Logtalk for VSCode plug-in published:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

You can now install the plug-in from VSCode by going to the Extensions tab and searching for "Logtalk".

**Reply by pmoura** [reply]
Also published in the VSCodium marketplace:

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.20.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

This is a major release, implementing code navigation features (requires Logtalk 3.78.0 or later version):

- Go to Declaration
- Go to Definition
- Go to Type Definition
- Go to References
- Go to Implementations
- Go to Symbol in Editor...
- Go to Symbol in Workspace...
- Show Call Hierarchy
- Show Type Hierarchy

There's also a new "Open Parent File" command. For details and a complete list of changes, please consult the release notes at:

https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.20.0

**Comment by pmoura** [comment]
Logtalk for VSCode 0.21.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

This is a major release, providing better integration with developer tools, new commands, improved usability, and bug fixes:

- Changed commands that run the developer tools to require the code to be loaded first
- The user is now warned when no code is loaded for a command that requires it
- The user is now informed when commands that spawn processes complete
- Added experimental code lens support for test results
- Added experimental code lens support for entity cyclomatic complexity
- Added "Logtalk: Compute Metrics" command
- Added "Logtalk: Toggle Code Lens" command
- Added "Logtalk: Generate Documentation (workspace)" command
- Added "Logtalk: Generate Diagrams (workspace)" command
- Added "Logtalk: Scan Dead Code (workspace)" command
-
Added "dead_code_scanner" tool warnings to the "Problems" pane
- Added "lgtdoc" tool warnings to the "Problems" pane
- Added "make" tool warnings to the "Problems" pane
- Added tests compilation warnings and errors to the "Problems" pane
- Added doclet compilation warnings and errors to the "Problems" pane
- Updated the "Known Issues" section in the readme file
- Fixed taking into account environment settings when spawning auxiliary Logtalk processes
- Fixed off-by-one error when parsing linter warnings lines
- Fixed deleting an atom or variable when typing an underscore before the first character

https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.21.0

**Comment by pmoura** [comment]
Logtalk for VSCode 0.22.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

This is a major release, providing improved usability, initial debugging support, new commands, and bug fixes:

* Change linters to no longer create and write to an "OUTPUT" pane channel
* Show current clause in the active text editor when debugging
* Add support for adding and removing spy points and log points
* Add Logtalk source file icons
* Add "Logtalk: Make - Circular" command
* Add "Logtalk: Make - Optimal" command
* Add "Logtalk: Make - Normal" command
* Add "Logtalk: Make - Debug" command
* Add "Logtalk: Make - Clean" command
* Add "Logtalk: Make - Caches" command
* Improve usability by only showing the terminal if hidden when advisable by the command
* Update minimum VSCode version required to 1.64.0
* Update the "Go to Implementations" command to also find protocol implementations
* Fix file recompilation to clear previous diagnostics for the file
* Fix possible JavaScript error in the "Go to Symbol in Editor..." implementation
* Fix occasional glitch where code navigation would return a previous result
* Fix and simplify auto-indentation patterns
* Fix parsing of predicate calls with double-quoted arguments

https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.22.0

**Comment by pmoura** [comment]
Logtalk for VSCode 0.23.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

This is a major release, providing improved usability, improved debugging support, new commands, and bug fixes:

* Rename workspace commands to include instead the word "Project"
* Show only project (workspace) commands in the explorer context menu
* Change the "Logtalk: Make ..." commands to show a warning when no Logtalk process is running
* Add "Logtalk: Create Project" command
* Add "Logtalk: Load Project" command
* Add getting started walkthrough
* Add failed tests to the "Problems" pane (when using the "Logtalk: Run Tests" command)
* Add support for conditional and triggered breakpoints
* Escape log point messages to ensure they are valid quoted atoms
* More informative error message when a script is not found
* Fix workspace commands when run from the command palette by using the first workspace folder
* Fix the "Logtalk: Load Directory" command to show a warning if the loader file doesn't exist
* Fix the "Logtalk: Run Tests" command to show a warning if the tester file doesn't exist
* Fix the "Logtalk: Run Doclet" command to show a warning if the doclet file doesn't exist

https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.23.0

**Comment by pmoura** [comment]
Logtalk for VSCode 0.26.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

Extension configuration is now significantly simplified. In most common cases, only the `LOGTALKHOME` and `LOGTALKUSER` environment variable absolute paths plus the Prolog backend identifier settings are required. There's also improved documentation and some bug fixes, notably on Windows.

**Comment by pmoura** [comment]
Logtalk for VSCode 0.29.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

Recent changes (since version 0.26.0):

- Label file tests and metrics data displayed using CodeLens as possibly outdated when editing the file
- Display tutor tool explanations in the "PROBLEMS" pane
- Fix reporting of failed tests when re-running tests
- Improve description of project (workspace) commands
- Fix code navigation issues when defining only the required settings

**Comment by pmoura** [comment]
Logtalk for VSCode 0.30.0 is now available:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

Main changes:

- Update readme section on code navigation features
- Fix code navigation issues on Windows
- Fix code navigation false warnings of code not loaded on Windows

**Comment by pmoura** [comment]
Logtalk for VSCode 0.32.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

Requires Logtalk 3.87.0 or a later version.

Recent changes (since 0.29.0):

- Add support for re-running a single test using CodeLens
- Fix case where duplicated items could be created in the "PROBLEMS" pane
- Fix "Logtalk: Toggle Code Lens" command to preserve code lens outdated status
- Fix code navigation issues on Windows
- Fix code navigation false warnings of code not loaded on Windows

**Comment by pmoura** [comment]
Logtalk for VSCode 0.33.0 is now available from both VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

Requires Logtalk 3.87.0 or a later version.

Changes in this version:

* Add "Logtalk: Jupyter" commands (requires Juyptext 1.16.7 or later version)
* Provide a menu with the valid choices for the backend setting instead of requiring typing it

**Comment by pmoura** [comment]
Logtalk for VSCode 0.42.0 released, featuring an experimental Copilot chat participant and a logger system. Available from both VSCode and VScodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.52.0 released with new features, usability improvements, and fixes (requires Logtalk 3.93.0):
 
* Update "Go to Symbol in Editor..." support to also include predicate clauses and non-terminal rules
* Add support for the "Run > Start Debugging" and "Run > Run Without Debugging" commands
* Add support for toggling debugging using a Logtalk icon in the top-right corner
* Improve readme section on requirements
* Fix spurious "No code loaded from selected directory as required by command." warning for the Logtalk built-in entities directory
* Fix spurious "No code loaded from selected directory as required by command." warning for loaded dependencies
* Fix cases where the number of arguments in a predicate call or entity identifier would be miscounted

Other recent noteworthy changes:

* Remove workaround of requiring selecting a predicate or entity name to enable the "Got to Definition" command
* Filter accidental code navigation calls with invalid predicate or entity indicators
* Fix possible file write permission errors when using the code navigation features
* Fix goal error when cancelling the creation of a function breakpoint
* Create a Logtalk terminal when the extension is first activated
* Show the Logtalk terminal also when loading a project, loading a directory, loading a file, and computing metrics
* Fix cases where the "Problems" pane would show duplicated diagnostics
* Remove code lens data for a source file when edited
* Improve automatic indentation support


Only available from the VSCode Marketplace for now. The Open VSX Registry (VSCodium) support for publishing extensions is currently down. But you can always download the .vsix file from https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode and install it manually (View > Command Palette... and then Extensions: Install from VSIX...).

**Comment by pmoura** [comment]
Logtalk for VSCode 0.53.0 released (requires Logtalk 3.94.0):

* Change caching of loaded directory paths to use an in-memory set to avoid workspace persistent state issues
* Add experimental support for the "Format Document" and "Format Selection" commands
* Add "Add argument to predicate/non-terminal" refactoring support
* Add "Reorder predicate/non-terminal arguments" refactoring support
* Add "Remove argument from predicate/non-terminal" refactoring support
* Add "Add parameter to object/category" refactoring support
* Add "Reorder object/category parameters" refactoring support
* Add "Remove parameter from object/category" refactoring support
* Add "Extract protocol" refactoring support
* Add "Extract to new Logtalk entity" refactoring support
* Add "Extract to new Logtalk file" refactoring support
* Add "Replace magic number with predicate call" refactoring support
* Add "Replace with include/1 directive" refactoring support
* Add "Replace include/1 directive with file contents" refactoring support
* Add entity, predicate, and non-terminal rename support
* Add "Logtalk: Compute Project Metrics" command
* Add settings support for multi-backend Logtalk executable arguments
* Add quick fixes for some linter errors and warnings
* Add quick fixes for some tests reporter warnings
* Add quick fixes for some dead code scanner warnings
* Add code navigation support for `(@)/1` goals
* Add "Logtalk: Make" submenu to the explorer context menu
* Add "logtalk.make.onSave" setting for automatically calling the "Logtalk: Make - Reload" command when saving a Logtalk source file
* Streamline presentation of symbols for the "Go to Symbol in Editor..." and "Go to Symbol in Workspace..." features
* Update "Go to Symbol in Workspace..." support to also include predicate clauses and non-terminal rules
* Improve performance of "Go to Symbol in Workspace..." and "Go to Symbol in Editor..." support
* Improve context menus items grouping and ordering
* Improve readme section on code navigation
* Improve extension logging for internal debugging
* Improve resource cleanup on extension deactivation
* Group settings into "Required", "Optional", and "Defaults" groups
* Set debugging off and ensure any existing breakpoints are disabled at extension activation
* Fix toggling debugging to always send the `debug`/`nodebug` messages after enabling/disabling the breakpoints
* Fix call hierarchy "Show Incoming Calls" and "Show Outgoing Calls" context menu items
* Fix "Go to Symbol in Editor..." and "Go to Symbol in Workspace..." support to parse multi-line predicate scope directives
* Fix "Go to Symbol in Editor..." and "Go to Symbol in Workspace..." support when entity, predicate, and non-terminal names contain single quotes
* Fix "Go to Symbol in Editor..." parsing of predicates and non-terminals with the same name but different arity
* Fix "Go to Symbol in Editor..." support when constructing predicate and non-terminal indicators from definitions with arguments using character code notation
* Fix code navigation commands to reject calls from inside line comments and support cancellation

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode
https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.64.0 released (requires Logtalk 3.95.0):

* Add "logtalk.diagrams.format" setting for specifying the diagrams format (default is `dot`; Graphviz)
* Add support for viewing SVG diagrams and HTML files in a webview with navigation and link handling
* Add "Logtalk: Open SVG in Viewer" command to open SVG files in a webview
* Add support for displaying code profiling results in a webview
* Add "Logtalk: Profiling" sub-menu with profiling commands
* Add support for the VS Code Testing API
* Add command "Logtalk: Run Tests with Coverage" to run all tests with coverage reporting
* Delete any existing `.vscode_*` temporary files at extension activation
* Changed test results summary displayed using CodeLens in the tests object to re-run only the object tests
* Add refactoring support for converting between object, protocol, and category entity types
* Add quick fixes for documentation linter warnings
* Additional quick fixes for compiler linter errors and warnings
* Additional quick fixes for the dead code scanner warnings
* Clear diagnostics from all diagnostic collections when a file is deleted or moved
* Clear diagnostics from all diagnostic collections when the Logtalk terminal is closed
* Provide limited support for virtual workspaces
* Fix running the documentation, diagrams, testing, and doclets scripts on Windows using default settings
* Fix diagnostics showing absolute paths instead of relative paths on Windows with some backends
* Fix test results CodeLens showing multiple occurrences of the `(outdated)` text after running individual tests
* Fix timing issue that could result in multiple Logtalk terminals being created at startup
* Fix bug where closing a non-Logtalk terminal would mark the Logtalk terminal as closed
* Fix the delete action for the dead code scanner to correctly check for predicate arity
* Fix predicate argument parsing bugs with commas and parentheses in quoted atoms
* Fix bug when computing a predicate definition range that would include clauses from other predicates
* Fix bugs handling diagnostic collections that could result in duplicated diagnostics after fixing and reloading files
* Fix errors with the debugging commands when the Logtalk terminal unexpectedly doesn't exist
* Fix walkthrough to prevent being closed when performing some of the steps
* Fix type hierarchy "Show Subtypes" and "Show Supertypes" context menu items
Available from both the VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode  
https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.75.0 released (requires Logtalk 3.96.0) with support for renaming local variables and parameter variables; new refactorings for extracting predicates, introducing and inlining variables, renumbering variables, wrapping plain Prolog files as objects, inferring public predicates, and sorting source files in driver files by dependencies; and improved performance of entity parameter refactorings.

Other recent changes in previous versions include fixes and improvements to the chat participant (including new slash commands); support for expanding and shrinking selections; updated documentation; propagation of file renames and deletions to loader and tester driver files with preview support; new refactorings for converting Prolog modules to objects, adding predicate/non-terminal declarations, splitting directives with list arguments, and moving code between entities; improved performance of the tests explorer and CodeLens providers; code completions when typing lists; creation of diagnostics from workspace tester and doclet  commands output; plus a new setting for loading the project on extension activation.

For details, see:

https://github.com/LogtalkDotOrg/logtalk-for-vscode/blob/master/CHANGELOG.md

Available from both the VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode  
https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.80.0 released (requires Logtalk 3.97.0) with support for the ECLiPSe backend on Windows; improved support for the SICStus Prolog backend; improved debugging with automatic code recompilation in debug/normal mode when starting/stopping debugging, support for the debug toolbar with additional buttons for common port commands, and partial support for the "Variables" and "Call Stack" panes in the "Run and Debug" sidebar; improved testing support with a new setting for automatically generating Allure reports, context menu items for skipping/unskipping tests, fixed clearing of failed tests diagnostics when editing and saving tests files, and reporting of automated test results in the Test Explorer sidebar; a new help command for opening the Logtalk Handbook in a view pane; multi-root workspace support; new code completions; a new refactoring for for replacing explicit message sending calls with a `uses/2` directive; actionable status bar indicators for profiling and CodeLens on/off state; improved handling of terminal crashes; improved code navigation and CodeLens performance; improved documentation; and fixed displaying of inline code coverage data on Windows.

Other recent changes in previous versions include new and improved refactorings; support for ctrl-click on file paths in "loaded" comment messages in the terminal; fixed settings grouping in the VS Code settings editor; and Windows fixes including handling of terminal links and a workaround for driver letter casing issues due to inconsistent path handling between Windows, VSCode, and Node.js APIs.

For details, see:

https://github.com/LogtalkDotOrg/logtalk-for-vscode/blob/master/CHANGELOG.md

Available from both the VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode  
https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.81.0 released (requires Logtalk 3.97.0). This is a minor release adding support for the GNU Prolog native code backend and fixing a Windows only issue with settings. For details, see:

https://github.com/LogtalkDotOrg/logtalk-for-vscode/blob/master/CHANGELOG.md

Available from both the VSCode and VSCodium marketplaces:

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode
https://open-vsx.org/extension/LogtalkDotOrg/logtalk-for-vscode

**Comment by pmoura** [comment]
Logtalk for VSCode 0.82.0 published:

https://github.com/LogtalkDotOrg/logtalk-for-vscode/releases/tag/v0.82.0

Available from both VSCode and VSCodium marketplaces.
